Provide a single place to create HTTP clients with different configurations

linukaratnayake opened the pull request:
## Purpose
Currently, when HTTP clients are needed they are created on the spot and used. This makes managing them difficult. Having one place to create and manage HTTP clients makes development easier.

## Goals
Create a new package `httpclient5` inside `carbon.utils` to include the classes and methods needed to create HTTP clients with different configurations.

## Approach
`httpclient5` package will include the classes and methods needed to create HTTP clients with different configurations. This class will act as the place that handles creation of HTTP clients. Apache HTTP Client 5 is used as the HTTP client, opposed to previously used older versions. 

## Related PRs
https://github.com/wso2/carbon-kernel/pull/4270

linukaratnayake committed:
Create copies of classes and modify

linukaratnayake committed:
Create whole new class for custom host name verifier

linukaratnayake committed:
Use DefaultHostnameVerifier underneath

linukaratnayake committed:
Bundle features

linukaratnayake committed:
Move custom verifier method to existing class

linukaratnayake committed:
Remove imports of Client 4

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.com).
+ * Copyright (c) 2023,2025, WSO2 LLC. (http://www.wso2.com).
```
```suggestion
 * Copyright (c) 2023-2025, WSO2 LLC. (http://www.wso2.com).
```

hwupathum commented on features/org.wso2.carbon.core.common.feature/pom.xml:
```diff
@@ -100,6 +100,8 @@
                                     org.wso2.orbit.commons-collections:commons-collections:${orbit.version.commons.collection}
                                 </bundleDef>
                                 <bundleDef>org.apache.httpcomponents.wso2:httpcore:${orbit.version.httpcore}</bundleDef>
+                                <bundleDef>org.wso2.orbit.org.apache.httpcomponents:httpcore5:${orbit.version.httpcore5}</bundleDef>
```
Can we remove  them here since these are already bundled to the IS pack

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/CustomHostNameVerifierNew.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Custom hostname verifier class.
+ */
+public class CustomHostNameVerifierNew implements HttpClientHostnameVerifier {
```
Lets use a different class name like `Http5CustomHostNameVerifier`

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -40,14 +44,45 @@ private HTTPClientUtils() {
+     *
+     * @return HttpClientBuilder.
+     */
+    public static HttpClientBuilder createClientWithCustomVerifierNew() {
```
Change method name to createHttp5ClientWithCustomVerifier

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -40,14 +44,45 @@ private HTTPClientUtils() {
      * @return HttpClientBuilder.
      */
-    public static HttpClientBuilder createClientWithCustomVerifier() {
+    public static org.apache.http.impl.client.HttpClientBuilder createClientWithCustomVerifier() {
```
Let's deprecate this method

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -40,14 +44,45 @@ private HTTPClientUtils() {
+
+        HttpClientBuilder httpClientBuilder = HttpClientBuilder.create().useSystemProperties();
+
+        HostnameVerifier hostnameVerifier = null;
```
Why do we use HostnameVerifier instead of X509HostnameVerifier?

linukaratnayake replied:
`X509HostnameVerifier` is an interface available in Apache Http Client 4. `HostnameVerifier` is in Apache Http Client 5 and it is the superinterface for both `HttpClientHostnameVerifier` and `NoopHostnameVerifier` used here.

linukaratnayake committed:
Change year in license header

Co-authored-by: Udara Pathum <46132469+hwupathum@users.noreply.github.com>

linukaratnayake committed:
Resolve PR comments - rename classes and methods

linukaratnayake committed:
Resolve PR comments

linukaratnayake committed:
Rewrite to use subject alternative names and common names similar to the original implementation

linukaratnayake committed:
Remove unwanted method

linukaratnayake committed:
Reformat file

linukaratnayake committed:
Reformat file - improve readability, add logging

linukaratnayake commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -38,16 +42,52 @@ private HTTPClientUtils() {
+
+        HostnameVerifier hostnameVerifier = null;
+        if (DEFAULT_AND_LOCALHOST.equals(System.getProperty(HOST_NAME_VERIFIER))) {
+            hostnameVerifier = new Http5CustomHostNameVerifier();
```
make a singleton

linukaratnayake commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/Http5CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,144 @@
+        
+        if (subjectAltNames != null) {
+            for (List<?> san : subjectAltNames) {
+                if (san != null && san.size() >= 2) {
```
Check instance type

linukaratnayake commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/Http5CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,144 @@
+                if (san != null && san.size() >= 2) {
+                    // DNS names are type 2, IP addresses are type 7.
+                    Integer type = (Integer) san.get(0);
+                    if (type == 2 || type == 7) {
```
Don't use ==

linukaratnayake commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/Http5CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,144 @@
+        X500Principal principal = cert.getSubjectX500Principal();
+        String dn = principal.getName(X500Principal.RFC2253);
+        
+        // Parse the DN to find CN.
```
check for `null`

linukaratnayake committed:
Make hostname verifiers singleton and type check subject alternative names

linukaratnayake committed:
Check distinguishedNames is null or empty

linukaratnayake committed:
Add method to create an Http Builder with trusting all certificates

linukaratnayake committed:
Add method to create client with system properties

linukaratnayake committed:
Create new package

linukaratnayake committed:
Rename variable for more clarity

linukaratnayake commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,148 @@
+                    if (typeObj instanceof Integer) {
+                        Integer type = (Integer) san.get(0);
+                        // DNS names are type 2, IP addresses are type 7.
+                        if (type.equals(2) || type.equals(7)) {
```
See whether only type 2 and 7 are used.

madurangasiriwardena commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/HTTPClientUtils.java:
```diff
@@ -38,8 +38,11 @@ private HTTPClientUtils() {
     /**
      * Get the httpclient builder with custom hostname verifier.
      *
+     * @deprecated This method uses the Apache HTTP Client 4.x. Use Apache HTTP Client 5.x instead.
```
This notice needs to be changed. First you need to say the replacement method and then explain the reason for deprecation if needed. Asking to use "Apache HTTP Client 5.x" does not make sense.

linukaratnayake replied:
Can you please tell the exact reason for the deprecation? According to my understanding, it is because, since anyway there is an ongoing unification effort, and all clients are going to be written using one library, why not use the latest one (5.x) instead of the older ones (4.x). Is this correct?

linukaratnayake committed:
Deprecate old CustomHostNameVerifier class

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,148 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(CustomHostNameVerifier.class);
+
+    private static final String[] LOCALHOST_ALTERNATIVES = {"::1", "127.0.0.1", "localhost", "localhost.localdomain"};
```
Lets rename this as localhost_san

hwupathum commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,148 @@
+ */
+public class CustomHostNameVerifier implements HttpClientHostnameVerifier {
+
+    public static final CustomHostNameVerifier INSTANCE = new CustomHostNameVerifier();
```
Use a getInstance() method for this

linukaratnayake committed:
Modify singleton declaration

madurangasiriwardena commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,156 @@
+/**
+ * Custom hostname verifier class with Apache Http Client 5.
+ */
+public class CustomHostNameVerifier implements HttpClientHostnameVerifier {
```
Have a better meaningful name for the class

madurangasiriwardena commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,156 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(CustomHostNameVerifier.class);
+
+    private static final String[] LOCALHOST_SAN = {"::1", "127.0.0.1", "localhost", "localhost.localdomain"};
```
```suggestion
    private static final String[] LOCALHOST_SANS = {"::1", "127.0.0.1", "localhost", "localhost.localdomain"};
```

madurangasiriwardena commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java:
```diff
@@ -0,0 +1,156 @@
+        List<String> result = new ArrayList<>();
+        
+        if (subjectAltNames != null) {
+            for (List<?> san : subjectAltNames) {
```
Add a comment explaining what's in the `san` list and why you have to take the 0th element.

linukaratnayake committed:
Explain why class is deprecated and how SANs collection is analyzed

linukaratnayake committed:
Explain why class is deprecated and how SANs collection is analyzed

linukaratnayake committed:
Add reason for deprecation

linukaratnayake committed:
Apply suggestions from code review

- Change how singleton pattern is used.
- Improve comments

Co-authored-by: Maduranga Siriwardena <maduranga.siriwardena@gmail.com>

linukaratnayake committed:
Rename method

Rename method to `createClientWithCustomHostnameVerifier` from `createClientWithCustomVerifier`.

Co-authored-by: Udara Pathum <46132469+hwupathum@users.noreply.github.com>

linukaratnayake committed:
Remove unwanted method

linukaratnayake committed:
Make singleton instance 'final'

Co-authored-by: Maduranga Siriwardena <maduranga.siriwardena@gmail.com>

linukaratnayake committed:
Fix PR comments

hwupathum commented:
Closing since https://github.com/wso2/carbon-kernel/pull/4302
